Parallel sync fix

roystgnr opened the pull request:
## Reason
Fixes #23796.  This fixes the hung parallel runs, and adds assertions to help make those runs a bit easier to debug in the future.

## Design
The first commit wraps `loopOverMortarSegments` in a `PARALLEL_CATCH`, to synchronize handling of any exceptions (like the quadrature point exception thrown in the test case) between ranks.

The second commit adds new assertions.

## Impact
The fix keeps asynchronous exceptions from hanging a whole job, at the cost of adding a `maxloc` that makes exceptionless runs infinitesimally slower.

The new assertions should give fewer hangs and more accurate error messages when any similar bugs are hit in the future, at the cost of adding `max`/`min` calls that make `devel` and `dbg` runs infinitesimally slower.

roystgnr committed:
Do loopOverMortarSegments with a PARALLEL_CATCH

Otherwise we can hit a quadrature error on just one processor and lose
parallel synchronization because of it.

This fixes #23796 for me.

roystgnr committed:
Add some parallel_only assertions

I still want to tear through MOOSE with these calls one day, get us more
sensible error messages and fewer hung processes when running in devel
and dbg modes, but for now here are the assertions I used while
debugging #23796

lindsayad commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -129,18 +131,29 @@ ComputeMortarFunctor::operator()(const Moose::ComputeType compute_type)
-                                        act_functor);
+  PARALLEL_TRY
+  {
+    try
+    {
+      Moose::Mortar::loopOverMortarSegments(iterators,
+                                            _assembly,
+                                            _subproblem,
+                                            _fe_problem,
+                                            _amg,
+                                            _displaced,
+                                            _mortar_constraints,
+                                            0,
+                                            _secondary_ip_sub_to_mats,
+                                            _primary_ip_sub_to_mats,
+                                            _secondary_boundary_mats,
+                                            act_functor);
+    }
+    catch (MooseException & e)
+    {
+      _fe_problem.setException(e.what());
+    }
```
So in `ThreadedElementLoopBase` we have a nested `try-catch` that catches `libMesh::LogicError` and `MetaPhysicL::LogicError` and re-throws as a `MooseException` which is caught by the outer `try-catch`. I think we should add that here as well. Also before setting the exception we should assert that we are not running threads (`ThreadedElementLoopBase`, because it is threaded, alternatively locks a mutex)

roystgnr replied:
That seems sane.  The double `catch(foo)` in `AuxilliarySystem::computeMortarNodalVars` has a cleaner look to it, though, for the same effect; I'll do that unless you object.

Actually ... is it sane *enough*?  Looking through our `PARALLEL_CATCH` usage elsewhere, I see `catch(MetaPhysicL::LogicError & e)` ... and for future-proofing there's no telling what we might have coming out of libraries some day.  I'd be tempted to `catch(...)` and get everything ... except that for *true* logic errors we really do want to just scream-and-die right away while there's a hope of debugging.  `parallel_object_only()` would be a perfect case in point.  It's just an artifact of past design mistakes that libMesh sees geometry errors and throws exceptions calling them logic errors, thus forcing app codes to treat `LogicError` as if it wasn't supposed to be just for bugs...

In my own code I'd used `libMesh::ConvergenceFailure` for "stuff you can catch and recover from", but that was in a world without tangled meshes, where it was just solver convergence issues forcing me to back off and retry.  I think the long run solution might be to create ... call it `libMesh::EvaluationFailure`, maybe? ... make it a subclass of `LogicError` at first for backwards compatibility, then switch the catch blocks to catch it directly, then break it out into just a subclass of `std::runtime_error`.  `ConvergenceFailure` would be a subclass of `EvaluationFailure`, too, to use for e.g. nested iterative solvers.

But that's long run.  Here I'll just make the suggested change, unless you think my immediate fears are right and we ought to be catching `MetaPhysicL::LogicError` too.

lindsayad replied:
Yes the catches at the same level look good too. And yes if you could add a catch for `MetaPhysicL::LogicError` in `AuxilliarySystem::computeMortarNodalVars` that would be much appreciated

roystgnr replied:
Plain catch and `setException(e.what())`, or `translateMetaPhysicLError`?  The latter does rethrow under the hood so we do need nested catches if you'd prefer that.

lindsayad reviewed (changes requested):
Thanks so much for ~begugging~ debugging this Roy

roystgnr committed:
Look for libMesh exceptions here too

roystgnr committed:
Use a nicer re-throw idiom

This preserves the underlying message, and it's used elsewhere in this
very file.

roystgnr commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -129,18 +131,33 @@ ComputeMortarFunctor::operator()(const Moose::ComputeType compute_type)
+                                            _secondary_boundary_mats,
+                                            act_functor);
+    }
+    catch (libMesh::LogicError & e)
+    {
+      _fe_problem.setException("We caught a libMesh::LogicError: " + std::string(e.what()));
+    }
+    catch (MooseException & e)
+    {
+      _fe_problem.setException(e.what());
+    }
```
Wow.  I would not have considered exceeding an array size to be a recoverable error ... though yeah, now I see it in this case.  Definitely need some more thought before I go hog wild on a "figure out how to recover from precisely the set of exceptions that are recoverable" design...

lindsayad commented:
Ok I have two threads talking about the same thing and I've contradicted myself between one conversation and what I've actually put into code. Looking at `translateMetaPhysicLError` ... I'm now undecided. @recuero do you have any sense of how often we have really oblique projections resulting in a really large AD container fill that can be saved by cutting the timestep and re-trying?

recuero commented:
That's a tough question to answer. 

 - Situations in which we have really oblique projections resulting in large required AD container size happen when we have close to perpendicular surfaces that come into contact. That sounds like a contradiction but it can easily happen when there is contact between a cube and a plane (e.g. with wrap-around sidesets) or a sphere and a plane. So, it's a pretty common situation.
 - Since oblique projections can happen in many cases, we should be able to cut the time step if something is off. 
 
 Can the large, required container size be mitigated by cutting the time step? I think that, in general, no. It could be that if an intermediate very distorted configuration generated very oblique projections and cutting the time step would resolve that particular issue, but my guess is that this occurrence would be very rare.

recuero commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -206,6 +206,8 @@ ComputeWeightedGapLMMechanicalContact::computeJacobian(const Moose::MortarType m
 void
 ComputeWeightedGapLMMechanicalContact::post()
 {
+  parallel_object_only();
```
We should also plan on extending this call to other mortar objects in the future, shouldn't we?

lindsayad replied:
yes

roystgnr replied:
IMHO there's probably a thousand places in MOOSE that would benefit from this call.  I figured for now I'd just leave this reactively-added batch in and see if there was much interest or much pushback before I considered getting proactive about it.

lindsayad commented:
Based off your response, my inkling would be to regard any time we exceed the configured AD container size as a hard error, and not something we should try to recover from.

In that vein, @roystgnr I would suggest we regard `MetaPhysicL::LogicError` as `mooseError`, at least until we have some distinction between "exceeded AD container size" types of errors and "I tried to take the log of a negative number" type of error. So I think we should use `moose::translateMetaPhysicLError` uniformly

roystgnr committed:
catch(MetaPhysicL) too

roystgnr commented:
```
tensor_mechanics/test:pressure.jacobian_RZ_quad8: /opt/moose/modules/tensor_mechanics/test/tests/pressure/ring.i:33: (Problem/coord_type):
tensor_mechanics/test:pressure.jacobian_RZ_quad8: The parameter 'coord_type' is deprecated.
tensor_mechanics/test:pressure.jacobian_RZ_quad8: Please use 'Mesh/coord_type' instead
tensor_mechanics/test:pressure.jacobian_RZ_quad8: 
tensor_mechanics/test:pressure.jacobian_RZ_quad8: 
tensor_mechanics/test:pressure.jacobian_RZ_quad8: ################################################################################
tensor_mechanics/test:pressure.jacobian_RZ_quad8: Tester failed, reason: EXPECTED OUTPUT NOT FOUND
```

In Docker/Test_modules?  I don't think that was me.

roystgnr commented:
> So I think we should use moose::translateMetaPhysicLError uniformly

Added; assuming I've interpreted you correctly.

> Because performing abort in one thread (which is what mooseError does) and not others can lead to very bad simulation tear-down.

I'd have hoped that `MPI_THREAD_FUNNELED` would give implementations enough of a warning to let `MPI_Abort` clean up after itself.  Digging into `timpi_init.C` to look into things brings up a couple paragraphs' comment about how many stacks aren't handling `MPI_THREAD_FUNNELED` requests properly, though, so maybe not.

lindsayad commented:
So you think conceptually `MPI_Abort` should be fine in a threaded region? I just recall working with @grmnptr on some problem where we were seeing use of a static mutex that had been destroyed already when we were aborting in a threaded region

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1422418/) on daa2dba wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23832/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1422432/) on daa2dba wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8f6c66</th>
      <th colspan=4>#23832 daa2db</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.89%</td>
      <td>84.89%</td>
      <td>-0.01%</td>
      <td>43.75%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84301</td>
      <td>84302</td>
      <td>+1</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15003</td>
      <td>15011</td>
      <td>+8</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23832/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23832/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23832/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 43.75% is less than the suggested 90.0%



This comment will be updated on new commits.
